GVFS/GVFS.Common/Maintenance/GitMaintenanceQueue.cs

```diff
@@ -0,0 +1,109 @@
+﻿using GVFS.Common.Tracing;
+using System;
+using System.Collections.Concurrent;
+using System.IO;
+using System.Threading;
+
+namespace GVFS.Common.Maintenance
+{
+    public class GitMaintenanceQueue
+    {
+        private GVFSContext context;
+        private BlockingCollection<GitMaintenanceStep> queue = new BlockingCollection<GitMaintenanceStep>();
+        private GitMaintenanceStep currentStep;
+
+        public GitMaintenanceQueue(GVFSContext context)
+        {
+            this.context = context;
+            Thread worker = new Thread(() => this.RunQueue());
+            worker.Name = "MaintenanceWorker";
+            worker.IsBackground = true;
+            worker.Start();
+        }
+
+        public void Enqueue(GitMaintenanceStep step)
+        {
+            try
+            {
+                this.queue?.Add(step);
+            }
+            catch (InvalidOperationException)
+            {
+                // We called queue.CompleteAdding()
+            }
+        }
+
+        public void Stop()
+        {
+            this.queue?.CompleteAdding();
+            this.currentStep?.Stop();
+        }
+
+        /// <summary>
+        /// This method is public for test purposes only.
+        /// </summary>
+        public bool EnlistmentRootReady()
+        {
+            // If a user locks their drive or disconnects an external drive while the mount process
+            // is running, then it will appear as if the directories below do not exist or throw
+            // a "Device is not ready" error.
+            try
+            {
+                return this.context.FileSystem.DirectoryExists(this.context.Enlistment.EnlistmentRoot)
+                         && this.context.FileSystem.DirectoryExists(this.context.Enlistment.GitObjectsRoot);
+            }
+            catch (IOException)
+            {
+                return false;
+            }
+        }
+
+        private void RunQueue()
+        {
+            while (true)
+            {
+                if (!this.queue.TryTake(out this.currentStep, Timeout.Infinite)
+                    || this.queue.IsAddingCompleted)
+                {
+                    // A stop was requested
+                    this.queue.Dispose();
+                    this.queue = null;
+                    return;
+                }
+
+                if (this.EnlistmentRootReady())

```

Comment:
We shouldn't do any retry logic. We will keep scheduling things and will catch up any missed work by running future work.

Reply:
So this will cause anything in the queue not to run and _hope_ that it will get scheduled again?  Maybe we need some comments around `EnqueueOneTimeStep` that there are no guarantees that it will run? Or maybe around the scheduler that it is a best effort?